File: scriptase/jobs/stage_actions.py
"""Step detail panel actions → existing engine run modes (step 2.4).

contracts.md §10 / product §18 / implementation-plan step 2.4.

Production step actions never introduce a second execution path. Each action
maps onto a ported run mode the canvas already uses:

  Run            → node_with_deps
  Test           → node_isolated
  Regenerate     → retry_failed   (single-node re-run; works for any status)
  Run From Here  → from_node

View Input / View Output / Provider / History / Approve are inspect or
checkpoint actions — not new run modes. Approve resumes a durable
``awaiting_approval`` pause via the engine (step 2.6).

``-P`` never appears in a stage name; provider capability is metadata only.
"""

from __future__ import annotations

from typing import Any, Mapping

from scriptase.jobs.stage_projection import (
    PRIMARY_STAGE_BY_TYPE,
    SIDE_BRANCH_STAGE_BY_TYPE,
    STAGE_BY_KEY,
    StageProjectionError,
    assign_nodes_to_stages,
)

# Stable action ids the Production UI and any future API share.
STAGE_ACTIONS = (
    "run",
    "test",
    "regenerate",
    "run_from_here",
    "view_input",
    "view_output",
    "provider",
    "history",
    "approve",
)

# Actions that start an engine run through POST /api/workflow/run.
EXECUTABLE_STAGE_ACTIONS = frozenset({
    "run",
    "test",
    "regenerate",
    "run_from_here",
})

# §18 action → ported run_mode. No new modes.
ACTION_RUN_MODES: dict[str, str] = {
    "run": "node_with_deps",
    "test": "node_isolated",
    "regenerate": "retry_failed",
    "run_from_here": "from_node",
}

# Inspect / checkpoint actions — no engine start.
INSPECT_STAGE_ACTIONS = frozenset({
    "view_input",
    "view_output",
    "provider",
    "history",
    "approve",
})


class StageActionError(RuntimeError):
    """Stage action could not be resolved (``STAGE_ACTION_INVALID``)."""

    def __init__(self, message: str, *, details: Any = None):
        super().__init__(message)
        self.code = "STAGE_ACTION_INVALID"
        self.message = message
        self.details = details


def is_executable_action(action: str) -> bool:
    return action in EXECUTABLE_STAGE_ACTIONS


def run_mode_for_action(action: str) -> str:
    """Return the engine run_mode for an executable stage action.

    Raises
    ------
    StageActionError
        When *action* is unknown or is an inspect/checkpoint action.
    """
    if action in INSPECT_STAGE_ACTIONS:
        raise StageActionError(
            f"Action {action!r} does not start a run",
            details={"action": action, "kind": "inspect"},
        )
    mode = ACTION_RUN_MODES.get(action)
    if mode is None:
        raise StageActionError(
            f"Unknown stage action: {action!r}",
            details={"action": action, "known": list(STAGE_ACTIONS)},
        )
    return mode


def stage_primary_target(
    stage: Mapping[str, Any],
    workflow: Mapping[str, Any],
) -> str:
    """Pick the node id a stage action targets.

    Prefer the first member whose type is a primary (non-side-branch) owner of
    that stage. Fall back to the first member id (stable topo order from the
    projection). Multi-node stages (e.g. Composer) must not target a collapsed
    side branch when a primary is present.
    """
    if not isinstance(stage, Mapping):
        raise StageActionError("stage must be an object")
    node_ids = stage.get("node_ids") or []
    if not isinstance(node_ids, list) or not node_ids:
        raise StageActionError(
            "Stage has no member nodes to target",
            details={"stage_key": stage.get("key")},
        )
    nodes_by_id: dict[str, dict] = {}
    for node in (workflow or {}).get("nodes") or []:
        if isinstance(node, Mapping) and isinstance(node.get("id"), str):
            nodes_by_id[node["id"]] = dict(node)

    stage_key = stage.get("key")
    for node_id in node_ids:
        if not isinstance(node_id, str) or not node_id:
            continue
        node = nodes_by_id.get(node_id) or {}
        type_key = str(node.get("type") or "")
        primary_stage = PRIMARY_STAGE_BY_TYPE.get(type_key)
        if primary_stage is not None and (
            stage_key is None or primary_stage == stage_key
        ):
            return node_id

    # Side-branch-only stage (unusual) or unknown types — first valid id.
    for node_id in node_ids:
        if isinstance(node_id, str) and node_id:
            return node_id
    raise StageActionError(
        "Stage has no valid target node id",
        details={"stage_key": stage.get("key"), "node_ids": node_ids},
    )


def build_stage_run_request(
    action: str,
    stage: Mapping[str, Any],
    workflow: Mapping[str, Any],
    *,
    workflow_id: str | None = None,
    force: bool = False,
) -> dict[str, Any]:
    """Build the POST /api/workflow/run body a canvas-initiated run would send.

    The Production view and the Workflow canvas must produce identical requests
    for the same action + target so execution records match field by field
    (volatile ids/timestamps excluded).
    """
    run_mode = run_mode_for_action(action)
    target = stage_primary_target(stage, workflow)
    body: dict[str, Any] = {
        "run_mode": run_mode,
        "target_node_ids": [target],
        "force": bool(force),
    }
    resolved_id = workflow_id or (
        workflow.get("workflow_id") if isinstance(workflow, Mapping) else None
    )
    if isinstance(resolved_id, str) and resolved_id:
        body["workflow_id"] = resolved_id
    else:
        body["workflow"] = workflow
    return body


def resolve_stage_from_projection(
    projection: Mapping[str, Any],
    stage_key: str,
) -> dict[str, Any]:
    """Look up a stage row by key inside a StageProjection document."""
    if not isinstance(projection, Mapping):
        raise StageActionError("projection must be an object")
    stages = projection.get("stages") or []
    if not isinstance(stages, list):
        raise StageActionError("projection.stages must be an array")
    for stage in stages:
        if isinstance(stage, Mapping) and stage.get("key") == stage_key:
            return dict(stage)
    raise StageActionError(
        f"Unknown stage key: {stage_key!r}",
        details={
            "stage_key": stage_key,
            "known": [
                s.get("key") for s in stages if isinstance(s, Mapping)
            ],
        },
    )


def stage_for_key(
    workflow: Mapping[str, Any],
    stage_key: str,
) -> dict[str, Any]:
    """Build a minimal stage dict for *stage_key* from a workflow graph.

    Used by tests and callers that have not yet loaded a full projection.
    """
    if stage_key not in STAGE_BY_KEY:
        raise StageActionError(
            f"Unknown stage key: {stage_key!r}",
            details={"stage_key": stage_key, "known": list(STAGE_BY_KEY)},
        )
    try:
        assignment = assign_nodes_to_stages(workflow)
    except StageProjectionError as exc:
        raise StageActionError(exc.message, details=exc.details) from exc
    member_ids = list(assignment.get(stage_key) or [])
    meta = STAGE_BY_KEY[stage_key]
    return {
        "key": stage_key,
        "label": meta["label"],
        "ordinal": meta["ordinal"],
        "node_ids": member_ids,
        "status": "idle",
        "provider_capable": False,
        "active_provider_instance_id": None,
        "artifacts": [],
        "issues": [],
    }


def action_requires_provider(
    action: str,
    stage: Mapping[str, Any],
) -> bool:
    """True when the UI should surface provider selection for this action.

    Provider UI appears only where the stage is provider-capable *and* the
    selected mode is an executable run that needs a provider (§6, §19).
    Inspect actions and non-provider stages never require one.
    """
    if action not in EXECUTABLE_STAGE_ACTIONS:
        return False
    return bool(stage.get("provider_capable"))


def normalize_execution_record_for_compare(record: Mapping[str, Any]) -> dict[str, Any]:
    """Strip volatile fields so two equivalent runs can be compared field-wise.

    Execution ids, project ids, timestamps, durations, fingerprints, and log
    timestamps differ between otherwise identical canvas and Production starts.
    Scope, run mode, node statuses, and output shapes must match.
    """
    if not isinstance(record, Mapping):
        raise StageActionError("execution record must be an object")

    def scrub_node(node: Mapping[str, Any]) -> dict[str, Any]:
        out = {
            "status": node.get("status"),
            "attempts": node.get("attempts"),
            "from_sample_data": bool(node.get("from_sample_data")),
            "resolved_inputs_summary": node.get("resolved_inputs_summary") or {},
            "outputs_summary": node.get("outputs_summary") or {},
            "artifact_refs": list(node.get("artifact_refs") or []),
            "error": _scrub_error(node.get("error")),
            "attempt_errors": [
                _scrub_error(err) for err in (node.get("attempt_errors") or [])
                if isinstance(err, Mapping)
            ],
            "cache": _scrub_cache(node.get("cache")),
        }
        return out

    nodes_in = record.get("nodes") or {}
    nodes_out: dict[str, Any] = {}
    if isinstance(nodes_in, Mapping):
        for node_id, node in nodes_in.items():
            if isinstance(node, Mapping):
                nodes_out[str(node_id)] = scrub_node(node)

    return {
        # workflow_id is assigned when a draft is snapshotted; two otherwise
        # identical starts get distinct ids, so it is not part of parity.
        "run_mode": record.get("run_mode"),
        "scope_node_ids": list(record.get("scope_node_ids") or []),
        "status": record.get("status"),
        "nodes": nodes_out,
        "schema_version": record.get("schema_version", 1),
    }


def _scrub_error(error: Any) -> dict[str, Any] | None:
    if not isinstance(error, Mapping):
        return None
    return {
        "code": error.get("code"),
        "message": error.get("message"),
    }


def _scrub_cache(cache: Any) -> dict[str, Any] | None:
    if not isinstance(cache, Mapping):
        return None
    return {
        "hit": bool(cache.get("hit")),
        "reason": cache.get("reason"),
    }


# Side-branch type lookup kept exported for tests that assert Composer does
# not target captions when assemble is present.
def is_side_branch_type(type_key: str) -> bool:
    return type_key in SIDE_BRANCH_STAGE_BY_TYPE

File: frontend/src/features/production/stageActions.js
/**
 * Step detail panel actions → existing engine run modes (step 2.4).
 *
 * Mirrors scriptase.jobs.stage_actions. Each executable action maps onto a
 * ported run mode the canvas already uses — no second execution path.
 *
 *   Run            → node_with_deps
 *   Test           → node_isolated
 *   Regenerate     → retry_failed
 *   Run From Here  → from_node
 *
 * View Input / View Output / Provider / History / Approve are inspect or
 * checkpoint actions, not run modes. Provider UI appears only when the stage
 * is provider_capable and the action is executable (§6, §19).
 */

/** @typedef {'run'|'test'|'regenerate'|'run_from_here'|'view_input'|'view_output'|'provider'|'history'|'approve'} StageActionId */

export const STAGE_ACTIONS = Object.freeze([
  'run',
  'test',
  'regenerate',
  'run_from_here',
  'view_input',
  'view_output',
  'provider',
  'history',
  'approve',
])

export const EXECUTABLE_STAGE_ACTIONS = Object.freeze([
  'run',
  'test',
  'regenerate',
  'run_from_here',
])

/** §18 action → ported run_mode. */
export const ACTION_RUN_MODES = Object.freeze({
  run: 'node_with_deps',
  test: 'node_isolated',
  regenerate: 'retry_failed',
  run_from_here: 'from_node',
})

export const ACTION_LABELS = Object.freeze({
  run: 'Run',
  test: 'Test',
  regenerate: 'Regenerate',
  run_from_here: 'Run From Here',
  view_input: 'View Input',
  view_output: 'View Output',
  provider: 'Provider',
  history: 'History',
  approve: 'Approve',
})

export const ACTION_HINTS = Object.freeze({
  run: 'Execute this stage with upstream dependencies resolved first.',
  test: 'Open the Test Node panel: isolate this stage’s primary node with the input picker. Never advances the Job.',
  regenerate: 'Re-run only this stage’s primary node.',
  run_from_here: 'Run this stage and eligible downstream stages.',
  view_input: 'Inspect resolved inputs recorded for the primary node.',
  view_output: 'Inspect outputs and artifact refs for the primary node.',
  provider: 'Active provider instance for this provider-capable stage.',
  history: 'Attempt history, side-by-side artifact version comparison (provider instance, seed, prompt revision), and the repairs the correction loop ran for this Job.',
  approve: 'Mark the stage accepted at a human checkpoint (durable at 2.6).',
})

/**
 * @param {string} action
 * @returns {boolean}
 */
export function isExecutableAction(action) {
  return EXECUTABLE_STAGE_ACTIONS.includes(action)
}

/**
 * @param {string} action
 * @returns {string}
 */
export function runModeForAction(action) {
  const mode = ACTION_RUN_MODES[action]
  if (!mode) {
    throw new Error(
      action in ACTION_LABELS
        ? `Action "${action}" does not start a run`
        : `Unknown stage action: ${action}`,
    )
  }
  return mode
}

/**
 * Prefer the first primary (non-side-branch) member; else the first id.
 * Node types that own a stage via PRIMARY_STAGE_BY_TYPE win over collapsed
 * side branches (captions/music → Composer).
 *
 * @param {{ node_ids?: string[], key?: string }} stage
 * @param {{ nodes?: Array<{ id: string, type?: string }> }} [workflow]
 * @param {Record<string, string>} [primaryStageByType]
 * @returns {string}
 */
export function stagePrimaryTarget(
  stage,
  workflow = {},
  primaryStageByType = null,
) {
  const nodeIds = Array.isArray(stage?.node_ids) ? stage.node_ids : []
  if (!nodeIds.length) {
    throw new Error('Stage has no member nodes to target')
  }
  const nodesById = Object.create(null)
  for (const node of workflow?.nodes || []) {
    if (node && typeof node.id === 'string') nodesById[node.id] = node
  }

  // Default primary map mirrors backend PRIMARY_STAGE_BY_TYPE for the common
  // media types. Callers may inject the live map when available.
  const primaryMap = primaryStageByType || DEFAULT_PRIMARY_STAGE_BY_TYPE
  const stageKey = stage?.key

  for (const nodeId of nodeIds) {
    if (typeof nodeId !== 'string' || !nodeId) continue
    const typeKey = String(nodesById[nodeId]?.type || '')
    const primaryStage = primaryMap[typeKey]
    if (primaryStage != null && (stageKey == null || primaryStage === stageKey)) {
      return nodeId
    }
  }
  for (const nodeId of nodeIds) {
    if (typeof nodeId === 'string' && nodeId) return nodeId
  }
  throw new Error('Stage has no valid target node id')
}

/**
 * Build the same POST /api/workflow/run body the canvas store sends.
 *
 * @param {StageActionId|string} action
 * @param {object} stage
 * @param {object} [options]
 * @param {string} [options.workflowId]
 * @param {object} [options.workflow]  draft document when no saved id
 * @param {boolean} [options.force]
 * @param {Record<string, string>} [options.primaryStageByType]
 * @returns {{ run_mode: string, target_node_ids: string[], force: boolean, workflow_id?: string, workflow?: object }}
 */
export function buildStageRunRequest(action, stage, options = {}) {
  const runMode = runModeForAction(action)
  const target = stagePrimaryTarget(
    stage,
    options.workflow || {},
    options.primaryStageByType || null,
  )
  const body = {
    run_mode: runMode,
    target_node_ids: [target],
    force: Boolean(options.force),
  }
  if (options.workflowId) {
    body.workflow_id = options.workflowId
  } else if (options.workflow) {
    body.workflow = options.workflow
  }
  return body
}

/**
 * @param {string} action
 * @param {{ provider_capable?: boolean }} stage
 * @returns {boolean}
 */
export function actionRequiresProvider(action, stage) {
  if (!isExecutableAction(action)) return false
  return Boolean(stage?.provider_capable)
}

/**
 * Whether Approve is currently meaningful for this stage.
 * Durable checkpoints land in step 2.6; until then only awaiting_approval.
 *
 * @param {{ status?: string }} stage
 * @returns {boolean}
 */
export function canApproveStage(stage) {
  return String(stage?.status || '') === 'awaiting_approval'
}

/** Keep in lockstep with scriptase.jobs.stage_projection.PRIMARY_STAGE_BY_TYPE */
export const DEFAULT_PRIMARY_STAGE_BY_TYPE = Object.freeze({
  'script.input': 'script',
  'story.generate': 'script',
  'tts.generate': 'voice',
  'timing.align': 'timing',
  'segment.run': 'segments',
  'scenes.blueprint': 'scenes',
  'storyboard.generate': 'images',
  'animator.generate': 'videos',
  'review.run': 'review',
  'review.validate': 'review',
  'review.semantic': 'review',
  'assemble.project': 'composer',
  'timeline.project': 'composer',
  'export.video': 'export',
})
